target/linux/ramips/dts/mt7621_iptime_t5004.dts

```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "mt7621.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+/ {
+	compatible = "iptime,t5004", "mediatek,mt7621-soc";
+	model = "ipTIME T5004";
+
+	aliases {
+		led-boot = &led_cpu;
+		led-failsafe = &led_cpu;
+		led-running = &led_cpu;
+		led-upgrade = &led_cpu;
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led_cpu: cpu {
+			label = "blue:cpu";
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+		};
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio 4 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+	};
+};
+
+&nand {
+	status = "okay";
+
+	partitions {
+		compatible = "fixed-partitions";
+		#address-cells = <1>;
+		#size-cells = <1>;
+
+		partition@0 {
+			label = "u-boot";
+			reg = <0x0 0x80000>;
+			read-only;
+		};
+
+		partition@80000 {
+			label = "config";
+			reg = <0x80000 0xc0000>;
+			read-only;
+		};
+
+		partition@140000 {
+			label = "firmware";
+			reg = <0x140000 0x7e40000>;
+
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "kernel";
+				reg = <0x0 0x400000>;
+			};
+
+			partition@400000 {
+				label = "ubi";
+				reg = <0x400000 0x7a40000>;
+			};
```

Comment:
This nesting of "firmware" and "ubi" partitions looks a bit odd to me. Are there other devices that use a similar scheme?

Reply:
```sh
$ pcregrep -Ml 'label(\\n|[^}])*"fixed-partitions"' target/linux/**/*.dts 
target/linux/ath79/dts/ar7241_ubnt_unifi-ap-outdoor-plus.dts
target/linux/ath79/dts/ar9344_ubnt_unifi-ap-pro.dts
target/linux/mvebu/files/arch/arm/boot/dts/armada-370-buffalo-ls421de.dts
target/linux/ramips/dts/mt7621_netis_wf2881.dts
```

Reply:
Hmm, only the last one is the valid one, and that node was written by me a few years ago...

Reply:
There are some devices with kernel partitions that overlap with rootfs partitions. Do you prefer it?
https://github.com/openwrt/openwrt/blob/master/target/linux/mvebu/files/arch/arm/boot/dts/armada-385-linksys-venom.dts#L154
https://github.com/openwrt/openwrt/blob/master/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-eax500.dtsi#L136

Reply:
Overlapping causes easier revert to stock and reflash when  soneone decide to change kernel partition size.

Reply:
I converted the partition table, but a partition named "kernel" overlapping with rootfs feels a bit uneasy to me...

Reply:
I tried searching again and found some more devices:
```sh
$ pcregrep -Ml '"firmware"(\n|.)*"ubi"' target/linux/**/*.dts{,i}
target/linux/ath79/dts/qca9558_zyxel_nbg6716.dts
target/linux/ramips/dts/mt7621_netis_wf2881.dts
target/linux/apm821xx/dts/netgear-wndap6x0.dtsi
$ pcregrep -Ml '"ubi"(\n|.)*"firmware"' target/linux/**/*.dts{,i}
target/linux/apm821xx/dts/netgear-wndr4700.dts
target/linux/ath79/dts/ar9344_netgear_wndr.dtsi
```
The "firmware" partition of these devices does not nest "kernel" and "ubi" partitions explicitly (by "fixed-partitions" compatible), but the range of "firmware" partition does include the "ubi" partition.
"firmware" and "kernel" partitions start at the same offset, so their nodes were named differently. Partition nesting scheme would not need this detour.

I still believe that "firmware" nesting "ubi" is not that odd or useless, it's just the counterpart of its NOR version, "firmware" nesting "rootfs".
What do you think? @adschm

Reply:
> target/linux/apm821xx/dts/netgear-wndr4700.dts

you are right, I should/will remove those overlapping firmware+dtb+nand nodes altogether.